zerver/lib/user_groups.py

```diff
@@ -226,20 +226,46 @@ def get_recursive_subgroups_for_groups(user_group_ids: List[int]) -> List[int]:
     return list(recursive_subgroups.values_list("id", flat=True))
 
 
+def get_role_based_system_groups_dict(realm: Realm) -> Dict[str, UserGroup]:
+    system_groups = UserGroup.objects.filter(realm=realm, is_system_group=True)
+    system_groups_name_dict = {}
+    for group in system_groups:
+        system_groups_name_dict[group.name] = group
+
+    return system_groups_name_dict
+
+
+def set_defaults_for_group_settings(user_group: UserGroup) -> UserGroup:
+    system_groups_name_dict = get_role_based_system_groups_dict(user_group.realm)
+
+    for setting_name, permission_config in UserGroup.GROUP_PERMISSION_SETTINGS.items():
+        default_group = system_groups_name_dict[permission_config.default_group_name]
+        setattr(user_group, setting_name, default_group)
```

Comment:
We can also pass a list of user groups also to this function and call `bulk_update` here instead of doing it in the caller of this function.

Reply:
That feels like it would complicate `create_user_group_in_database` more.

Reply:
Yes true.

Reply:
Yes true.

Reply:
Yeah this pattern seems mostly fine to me; the only database queries we do are in `get_role_based_system_groups_dict`.

I guess we're not calling this from `create_system_user_groups_for_realm`, but if we were to, we might want to refactor this in some way to avoid doing `get_role_based_system_groups_dict` in a loop.

Reply:
I did this refactoring to call `get_role_based_system_groups_dict` in this PR itself.